### PR TITLE
sql+: rust-dec-backed numeric type

### PR DESCRIPTION
**Pull request by sploiselle**
A very preliminary implementation of a numeric type (`rdn` for "rust-dec numeric") that supports to/from text, addition, and division. Want to use this as a conversation starter with looking at how we can do aggregations with `rust-dec`.

Note that there are some things in here that are a little YOLO because they don't tie into anything else, e.g. I've largely cargo-culted the Avro decimal work but didn't do much to test it; that stuff will all come later after we get a preliminary type working. The `adt::rdn` package also has an interface that's maybe too cute, but I won't know what that should look like until we build out more functions.

@umanwizard Would you mind taking a look at the Avro stuff to see if it makes sense at a high level?

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/6129)

### CODE REVIEW

**Comment by umanwizard** [comment]
What's the intent with the Avro decoding stuff. Will we be adding a new schema node type (e.g. `SchemaNode::RDN`) ? Without being driven by the schema, all of this will be dead code.

**Comment by sploiselle** [comment]
@umanwizard There will be a point where we swap over from using `Decimal` to `RDN` (though it will be called `Decimal` at that point`; so it will be dead code until that point. I included just because I wanted to make sure we had what we needed exposed in the underlying library.

File: src/repr/src/adt/rdn.rs
Comment (benesch) [inline src/repr/src/adt/rdn.rs:33]: Does `i128::from_be_bytes` not let you skip the `rev`?

File: src/pgrepr/src/value/rdn.rs
Comment (benesch) [inline src/pgrepr/src/value/rdn.rs:89]: Looks stray?

**Review by benesch — APPROVED** [review]
Woohoo! LGTM pending Brennan's review of the Avro bits.

**Comment by benesch** [comment]
Oh, btw, s/RDN/Rdn/ if you care, but don't worry about it because s/RDN/Numeric/ will happen eventually, yeah?

File: src/interchange/src/avro/decode.rs
Comment (umanwizard) [inline src/interchange/src/avro/decode.rs:467]: What do we intend to happen if someone has an Avro schema with a decimal field where `precision` is greater than 16?

**Comment by sploiselle** [comment]
@benesch You have any insight into this licensing failure? https://buildkite.com/materialize/tests/builds/15724#fb25d1d5-c348-4a25-ac86-06e07c5e2816

**Comment by benesch** [comment]
Ah yeah, ICU license just isn't allowlisted yet. Just drop it in deny.toml! It's OSI approved so we're good to use it.

File: src/interchange/src/avro/decode.rs
Comment (sploiselle) [reply src/interchange/src/avro/decode.rs:468]: The max precision is 34, but the idea here was just to treat it as if it were ~= an overflow and disallow it from being input. lmk if that makes sense!

Comment (umanwizard) [reply src/interchange/src/avro/decode.rs:468]: Well, it's a regression - it will mean we can't ingest avro sources with schemas that we currently can. Sorry, I'm not up to speed on the discussions about decimal types so perhaps the issues the new type is fixing is important enough that we're OK with just telling people with precision > 34 "sorry, you can't use Materialize."

Anyway, if we _do_ have a hard requirement of max precision 34, we should probably make sure to catch it at source creation time when we parse schemas, in addition to catching it here.

Comment (frankmcsherry) [reply src/interchange/src/avro/decode.rs:468]: > The max precision is 34

Is this not a regression from the current design?

Comment (sploiselle) [reply src/interchange/src/avro/decode.rs:468]: Just for those following along at home, this is a regression, but the current implementation's guarantee of 38 digits of precision doesn't actually hold up in some cases, e.g.

```
SELECT 1::NUMERIC(38, 10) / 1.2103500000::numeric(38, 10);
----
0.826000000000
```

The correct answer to this query should be ~`0.82620729541`, so we're offering 38 digits of precision, but then ~losing 7 at the end there, whereas this new library is capable of producing the correct result.

**Comment by philip-stoev** [comment]
@sploiselle  do you want me to test this at this stage? I see that it is possible to trigger computations using the new type from SQL.

**Comment by sploiselle** [comment]
@philip-stoev tysm for checking! not ready for testing at this stage, but will let you know once we have something worth poking at

**Pull request merged**
